From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Sat, 11 Jul 2026 20:16:20 +0800
Subject: [PATCH 33/47] Add abstraction for reading from main storage

These functions are intended for m1n1 to find the next stage on the boot
device.

On A7/A8/A8X: read from ANS1 storage

on A11+/Apple Silicon Macs: read from platform nvme

If m1n1 gains external boot support in the future, main storage may refer
to the external boot device

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 proxyclient/m1n1/proxy.py | 22 +++++++++++
 src/proxy.c               | 12 ++++++
 src/proxy.h               |  4 ++
 src/utils.c               | 77 ++++++++++++++++++++++++++++++++++++++-
 src/utils.h               |  4 ++
 5 files changed, 118 insertions(+), 1 deletion(-)

diff --git a/proxyclient/m1n1/proxy.py b/proxyclient/m1n1/proxy.py
index 53291d30..7a22d5e1 100644
--- a/proxyclient/m1n1/proxy.py
+++ b/proxyclient/m1n1/proxy.py
@@ -679,6 +679,14 @@ class M1N1Proxy(Reloadable):
     P_NVME_SHUTDOWN = 0xf01
     P_NVME_READ = 0xf02
     P_NVME_FLUSH = 0xf03
+    P_ANS1_INIT = 0xf04
+    P_ANS1_SHUTDOWN = 0xf05
+    P_ANS1_READ = 0xf06
+    P_ANS1_RESERVED = 0xf07
+    P_MAIN_STORAGE_INIT = 0xf08
+    P_MAIN_STORAGE_SHUTDOWN = 0xf09
+    P_MAIN_STORAGE_READ = 0xf0a
+    P_MAIN_STORAGE_FLUSH = 0xf0b
 
     P_MCC_GET_CARVEOUTS = 0x1000
 
@@ -1184,6 +1192,20 @@ class M1N1Proxy(Reloadable):
         return self.request(self.P_NVME_READ, nsid, lba, bfr)
     def nvme_flush(self, nsid):
         return self.request(self.P_NVME_FLUSH, nsid)
+    def main_storage_init(self):
+        return self.request(self.P_MAIN_STORAGE_INIT)
+    def main_storage_shutdown(self):
+        return self.request(self.P_MAIN_STORAGE_SHUTDOWN)
+    def main_storage_read(self, lba, bfr):
+        return self.request(self.P_MAIN_STORAGE_READ, lba, bfr)
+    def main_storage_flush(self):
+        return self.request(self.P_MAIN_STORAGE_FLUSH)
+    def ans1_init(self):
+        return self.request(self.P_ANS1_INIT)
+    def ans1_shutdown(self):
+        return self.request(self.P_ANS1_SHUTDOWN)
+    def ans1_read(self, lba, bfr):
+        return self.request(self.P_ANS1_READ, lba, bfr)
 
     def mcc_get_carveouts(self):
         return self.request(self.P_MCC_GET_CARVEOUTS)
diff --git a/src/proxy.c b/src/proxy.c
index 2471a89b..ff6d6e53 100644
--- a/src/proxy.c
+++ b/src/proxy.c
@@ -609,6 +609,18 @@ int proxy_process(ProxyRequest *request, ProxyReply *reply)
         case P_ANS1_RESERVED:
             reply->status = S_BADCMD;
             break;
+        case P_MAIN_STORAGE_INIT:
+            reply->retval = main_storage_init();
+            break;
+        case P_MAIN_STORAGE_SHUTDOWN:
+            main_storage_shutdown();
+            break;
+        case P_MAIN_STORAGE_READ:
+            reply->retval = main_storage_read(request->args[0], (void *)request->args[1]);
+            break;
+        case P_MAIN_STORAGE_FLUSH:
+            reply->retval = main_storage_flush();
+            break;
 
         case P_MCC_GET_CARVEOUTS:
             reply->retval = (u64)mcc_carveouts;
diff --git a/src/proxy.h b/src/proxy.h
index bb5a50f0..965807a4 100644
--- a/src/proxy.h
+++ b/src/proxy.h
@@ -168,6 +168,10 @@ typedef enum {
     P_ANS1_SHUTDOWN,
     P_ANS1_READ,
     P_ANS1_RESERVED,
+    P_MAIN_STORAGE_INIT,
+    P_MAIN_STORAGE_SHUTDOWN,
+    P_MAIN_STORAGE_READ,
+    P_MAIN_STORAGE_FLUSH,
 
     P_MCC_GET_CARVEOUTS = 0x1000,
 
diff --git a/src/utils.c b/src/utils.c
index cd3e0ecd..feeca692 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -3,16 +3,24 @@
 #include <assert.h>
 #include <stdarg.h>
 
-#include "utils.h"
+#include "adt.h"
+#include "ans1.h"
 #include "cpu_regs.h"
 #include "iodev.h"
+#include "nvme.h"
 #include "smp.h"
 #include "types.h"
 #include "utils.h"
 #include "vsprintf.h"
 #include "xnuboot.h"
 
+enum storage_type {
+    STORAGE_TYPE_ANS1 = 1,
+    STORAGE_TYPE_ANS23,
+};
+
 bool is_mac;
+static enum storage_type storage_type;
 
 static char ascii(char s)
 {
@@ -291,3 +299,70 @@ void cpu_sleep(bool deep)
         sysop("wfi");
     }
 }
+
+bool main_storage_init(void)
+{
+    int ans_node = adt_path_offset(adt, "/arm-io/ans");
+
+    if (ans_node < 0) {
+        /* A9/A9X/A10/A10X */
+        printf("Storage: storage not supported!\n");
+        return false;
+    }
+
+    if (adt_is_compatible(adt, ans_node, "iop,s5l8960x")) {
+        storage_type = STORAGE_TYPE_ANS1;
+        return ans1_init();
+    } else {
+        storage_type = STORAGE_TYPE_ANS23;
+        return nvme_init();
+    }
+}
+
+bool main_storage_read(u64 lba, u64 *bfr)
+{
+    if (!storage_type)
+        return false;
+
+    switch (storage_type) {
+        case STORAGE_TYPE_ANS1:
+            return ans1_read_main_storage(lba, bfr);
+        case STORAGE_TYPE_ANS23:
+            return nvme_read(1, lba, bfr);
+        default:
+            __builtin_unreachable();
+    }
+}
+
+void main_storage_shutdown(void)
+{
+    if (!storage_type)
+        return;
+
+    switch (storage_type) {
+        case STORAGE_TYPE_ANS1:
+            ans1_shutdown();
+            break;
+        case STORAGE_TYPE_ANS23:
+            nvme_shutdown();
+            break;
+        default:
+            __builtin_unreachable();
+    }
+}
+
+bool main_storage_flush(void)
+{
+    if (!storage_type)
+        return false;
+
+    switch (storage_type) {
+        case STORAGE_TYPE_ANS1:
+            /* XXX support unknown on ANS1 */
+            return true;
+        case STORAGE_TYPE_ANS23:
+            return nvme_flush(1);
+        default:
+            __builtin_unreachable();
+    }
+}
diff --git a/src/utils.h b/src/utils.h
index 94223e8b..429db16e 100644
--- a/src/utils.h
+++ b/src/utils.h
@@ -529,4 +529,8 @@ bool supports_gxf(void);
 bool supports_pan(void);
 u64 top_of_memory_alloc(size_t size);
 
+bool main_storage_init(void);
+bool main_storage_read(u64 lba, u64 *bfr);
+void main_storage_shutdown(void);
+bool main_storage_flush(void);
 #endif
